lectures/zksnark.tex

```diff
@@ -0,0 +1,812 @@
+
+So, we need to change the way how we manage coefficients. 
+
+\begin{definition} 
+    Consider a Rank-1 Constraint System (R1CS) defined by $m$ constraints. Each constraint is
+    associated with coefficient vectors $a_i b_i$ and $c_i$, where $i \in \{1, 2, \dots, m\}$ and
+    also a witness vector $w$ consisting of $n$ elements.
+
+    Then this system can also be represented using the corresponding matrices $A, B,$ and $C$.
+    \begin{align*}
+        A = \begin{pmatrix}
+            a_{11} & a_{12} & \dots & a_{1n} \\
+            a_{21} & a_{22} & \dots & a_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            a_{m1} & a_{m2} & \dots & a_{mn}
+        \end{pmatrix} & \quad
+        B = \begin{pmatrix}
+            b_{11} & b_{12} & \dots & b_{1n} \\
+            b_{21} & b_{22} & \dots & b_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            b_{m1} & b_{m2} & \dots & b_{mn}
+        \end{pmatrix} & 
+        C = \begin{pmatrix}
+            c_{11} & c_{12} & \dots & c_{1n} \\
+            c_{21} & c_{22} & \dots & c_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            c_{m1} & c_{m2} & \dots & c_{mn}
+        \end{pmatrix}
+    \end{align*}
+    
+    In this representation:
+    \begin{itemize}
+        \item Each row of the matrices corresponds to the coefficients of a specific constraint.
+        \item Each column of these matrices corresponds to the coefficients associated with a 
+        particular element of the witness vector $w$.
+    \end{itemize}
+\end{definition}
+
+\begin{example}
+    The vectors $a_i$ from the previous examples:
+    \begin{align*}
+        a_1 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_2 &= (0, 0, 0, 1, 0, 0, 0) \\
+        a_3 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_4 &= (1, 0, -1, 0, 0, 0, 0)
+    \end{align*}
+    $n = 7, m = 4$
+    \begin{equation*}
+        A = \begin{pmatrix}
+            a_{11} & a_{12} & a_{13} & a_{14} & a_{15} & a_{16} & a_{17} \\
+            a_{21} & a_{22} & a_{23} & a_{24} & a_{25} & a_{26} & a_{27} \\
+            a_{31} & a_{32} & a_{33} & a_{34} & a_{35} & a_{36} & a_{37} \\
+            a_{41} & a_{42} & a_{43} & a_{44} & a_{45} & a_{46} & a_{47}
+        \end{pmatrix} = \begin{pmatrix}
+            0 & 0 & 1 & 0 & 0 & 0 & 0 \\
+            0 & 0 & 0 & 1 & 0 & 0 & 0 \\
+            0 & 0 & 1 & 0 & 0 & 0 & 0 \\
+            1 & 0 & -1 & 0 & 0 & 0 & 0 
+        \end{pmatrix}
+    \end{equation*}
+\end{example}
+
+The columns of these matrices represent the mappings from constraint number $i$ to the corresponding
+coefficient of the $j$ element in the witness vector.
+
+\begin{example}
+    Considering the witness from the previos examples: 
+    \begin{equation*}
+         w = (1, r, x_1, x_2, x_3, \mathsf{mult}, \mathsf{selectMult})
+    \end{equation*}
+    For element $x_1$ we are interesting in the third columns of the $A$, $B$ and $C$ matrices, as
```

Comment:
For element $x_1$, we are interest**ed** in the third columns of $A,B,C$ matrices